### PR TITLE
ytmdl: init at 2021.05.26

### PR DESCRIPTION
**Pull request by j0hax**
###### Motivation for this change

Include the ytmdl package: _A simple app to get songs from YouTube in mp3 format with artist name, album name etc from sources like iTunes, LastFM, Deezer, Gaana etc._

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 127000` at 126132f8 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>11 packages built successfully:</summary>
  <ul>
    <li>python38Packages.downloader-cli</li>
    <li>python38Packages.itunespy</li>
    <li>python38Packages.pydes</li>
    <li>python38Packages.simber</li>
    <li>python38Packages.youtube_search</li>
    <li>python39Packages.downloader-cli</li>
    <li>python39Packages.itunespy</li>
    <li>python39Packages.pydes</li>
    <li>python39Packages.simber</li>
    <li>python39Packages.youtube_search</li>
    <li>ytmdl</li>
  </ul>
</details>
<details>
  <summary>5 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md">no-python-tests</a>

Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
Near pkgs/development/python-modules/pydes/default.nix:14:0:
```
   |
14 |     description = "A pure python module which implements the DES and Triple-DES encryption algorithms";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md">no-python-tests</a>

Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
Near pkgs/development/python-modules/simber/default.nix:16:0:
```
   |
16 |     description = "Simple, minimal and powerful logger for Python";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md">no-python-tests</a>

Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
Near pkgs/development/python-modules/itunespy/default.nix:16:0:
```
   |
16 |     description = "A simple library to fetch data from the iTunes Store API";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md">no-python-tests</a>

Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
Near pkgs/development/python-modules/downloader-cli/default.nix:16:0:
```
   |
16 |     description = "A simple downloader written in Python with an awesome customizable progressbar. ";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md">python-include-tests</a>

Consider adding a `checkPhase` for tests, or if not feasible, `pythonImportsCheck`.

Near pkgs/development/python-modules/youtube_search/default.nix:18:0:
```
   |
18 |     description = "Tool for searching for youtube videos to avoid using their heavily rate-limited API";
   | ^
```
</li>
  </ul>
</details>

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -9198,6 +9206,8 @@ in {
     phantomjsSupport = false;
   };
 
+  youtube_search = callPackage ../development/python-modules/youtube_search { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/python-packages.nix:9209]:
```suggestion
  youtube-search = callPackage ../development/python-modules/youtube-search { };
```

File: pkgs/tools/misc/ytmdl/default.nix [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    sha256 = "sha256-AYBhPmv1/cOaMh44kqm7JpJSudHbfriiJztcGL5mbEk=";
+  };
+
+  prePatch = ''
+    substituteInPlace setup.py \
+      --replace "bs4" "beautifulsoup4"
+  '';
+
+
+  propagatedBuildInputs = [ ffmpeg ffmpeg-python musicbrainzngs rich simber pydes youtube_search unidecode pyxdg downloader-cli beautifulsoup4 itunespy mutagen pysocks youtube-dl ];
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/ytmdl/default.nix:21]:
```suggestion
  propagatedBuildInputs = [ ffmpeg ffmpeg-python musicbrainzngs rich simber pydes youtube-search
  unidecode pyxdg downloader-cli beautifulsoup4 itunespy mutagen pysocks youtube-dl ];
```

File: pkgs/tools/misc/ytmdl/default.nix [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  version = "2021.05.26";
+
+  src = fetchFromGitHub {
+    owner = "deepjyoti30";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-AYBhPmv1/cOaMh44kqm7JpJSudHbfriiJztcGL5mbEk=";
+  };
+
+  prePatch = ''
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/ytmdl/default.nix:15]:
```suggestion
  postPatch = ''
```

File: pkgs/tools/misc/ytmdl/default.nix [hunk 4]
```diff
@@ -0,0 +1,32 @@
+{ lib, fetchFromGitHub, buildPythonApplication, ffmpeg, ffmpeg-python, musicbrainzngs, rich, simber, pydes, youtube_search, unidecode, pyxdg, downloader-cli, beautifulsoup4, itunespy, mutagen, pysocks, youtube-dl }:
+
+buildPythonApplication rec {
+
+  pname = "ytmdl";
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/ytmdl/default.nix:5]:
```suggestion
buildPythonApplication rec {
  pname = "ytmdl";
```

File: pkgs/development/python-modules/youtube_search/default.nix [hunk 5]
```diff
@@ -0,0 +1,23 @@
+{ lib, buildPythonPackage, fetchFromGitHub, requests }:
+buildPythonPackage rec {
+  pname = "youtube_search";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/youtube_search/default.nix:3]:
```suggestion
{ lib, buildPythonPackage, fetchFromGitHub, requests }:

buildPythonPackage rec {
  pname = "youtube-search";
```

File: pkgs/development/python-modules/youtube_search/default.nix [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  version = "unstable-2021-02-27";
+
+  src = fetchFromGitHub {
+    owner = "joetats";
+    repo = pname;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/youtube_search/default.nix:8]:
```suggestion
    repo = "youtube_search";
```

File: pkgs/development/python-modules/youtube_search/default.nix [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    rev = "886fe1b16c829215ee0984b6859f874b4a30d875";
+    sha256 = "sha256-3ECJ6iHNzx5PLgpTFraFzAYbKnyMYRf/iJ0zajU+hlo=";
+  };
+
+  propagatedBuildInputs = [ requests ];
+
+  doCheck = false;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/youtube_search/default.nix:15]: Why did you disable tests? Please add a comment.

Also pythonImportsCheck missing.

File: pkgs/development/python-modules/simber/default.nix [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  version = "0.2.2";
+
+  src = fetchFromGitHub {
+    owner = "deepjyoti30";
+    repo = pname;
+    rev = version;
+    sha256 = "0ksc2m61j5ijj0sq6kkc0hhkmfy9f51h9z3cl2sf8g6wbr9vc47h";
+  };
+
+  propagatedBuildInputs = [ colorama ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/simber/default.nix:13]: checkInputs and pythonImportsChecks?

File: pkgs/development/python-modules/pydes/default.nix [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  pname = "pydes";
+  version = "unstable-2019-01-08";
+
+  src = fetchFromGitHub {
+    owner = "twhiteman";
+    repo = "pyDes";
+    rev = "e988a5ffc9abb8010fc75dba54904d1c5dbe83db";
+    sha256 = "0sic8wbyk5azb4d4m6zbc96lfqcw8s2pzcv9nric5yqc751613ww";
+  };
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pydes/default.nix:12]: checkInputs and pythonImportsChecks?

File: pkgs/development/python-modules/pydes/default.nix [hunk 10]
```diff
@@ -0,0 +1,19 @@
+{ lib, buildPythonPackage, fetchFromGitHub }:
+buildPythonPackage rec {
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pydes/default.nix:2]:
```suggestion
{ lib, buildPythonPackage, fetchFromGitHub }:

buildPythonPackage rec {
```

File: pkgs/development/python-modules/itunespy/default.nix [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+  src = fetchFromGitHub {
+    owner = "sleepyfran";
+    repo = pname;
+    rev = version;
+    sha256 = "0yc3az5531qs8nbcw4rhgrszwczgy4bikfwfar7xb2044360sslw";
+  };
+
+  propagatedBuildInputs = [ requests pycountry mypy ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/itunespy/default.nix:14]: checkInputs and pythonImportsChecks?

File: pkgs/development/python-modules/itunespy/default.nix [hunk 12]
```diff
@@ -0,0 +1,21 @@
+{ lib, buildPythonPackage, fetchFromGitHub, requests, pycountry, mypy }:
+buildPythonPackage rec {
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/itunespy/default.nix:2]:
```suggestion
{ lib, buildPythonPackage, fetchFromGitHub, requests, pycountry, mypy }:

buildPythonPackage rec {
```

File: pkgs/development/python-modules/downloader-cli/default.nix [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+  version = "0.3.1";
+
+  src = fetchFromGitHub {
+    owner = "deepjyoti30";
+    repo = pname;
+    rev = version;
+    sha256 = "0gbbjxb9vf5g890cls3mwzl8lmcn6jkpgm5cbrif740mn2b4q228";
+  };
+
+  propagatedBuildInputs = [ urllib3 ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/downloader-cli/default.nix:13]: checkInputs and pythonImportsChecks?

File: pkgs/development/python-modules/downloader-cli/default.nix [hunk 14]
```diff
@@ -0,0 +1,21 @@
+{ lib, buildPythonPackage, fetchFromGitHub, urllib3 }:
+buildPythonPackage rec {
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/downloader-cli/default.nix:2]:
```suggestion
{ lib, buildPythonPackage, fetchFromGitHub, urllib3 }:

buildPythonPackage rec {
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/youtube_search/default.nix (hunk 7 above)
Comment (j0hax) [reply pkgs/development/python-modules/youtube_search/default.nix:15]: Hi, thank you for the reviews. I decided to disable checks as I get the following error when building without:

```python
FAILED (errors=1)
Test failed: <unittest.runner.TextTestResult run=1 errors=1 failures=0>
error: Test failed: <unittest.runner.TextTestResult run=1 errors=1 failures=0>
error: builder for '/nix/store/z0skj0gp3gn5ym0gwiak5c94y1vh4mm0-python3.8-youtube_search-unstable-2021-02-27.drv' failed with exit code 1;
       last 10 log lines:
       >     from .youtube_search import YoutubeSearch
       > ImportError: attempted relative import with no known parent package
       >
       >
       > ----------------------------------------------------------------------
       > Ran 1 test in 0.001s
       >
       > FAILED (errors=1)
       > Test failed: <unittest.runner.TextTestResult run=1 errors=1 failures=0>
       > error: Test failed: <unittest.runner.TextTestResult run=1 errors=1 failures=0>
       For full logs, run 'nix log /nix/store/z0skj0gp3gn5ym0gwiak5c94y1vh4mm0-python3.8-youtube_search-unstable-2021-02-27.drv'.
error: 1 dependencies of derivation '/nix/store/n63hkwgw5jm340zfsxrdb8vij9m3f82q-ytmdl-2021.05.26.drv' failed to build
```

Is there an option or patch I should be using to mitigate this error?

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 127000` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
11 packages built:
</summary>
<ul>
<li>python38Packages.downloader-cli</li>
<li>python38Packages.itunespy</li>
<li>python38Packages.pydes</li>
<li>python38Packages.simber</li>
<li>python38Packages.youtube-search</li>
<li>python39Packages.downloader-cli</li>
<li>python39Packages.itunespy</li>
<li>python39Packages.pydes</li>
<li>python39Packages.simber</li>
<li>python39Packages.youtube-search</li>
<li>ytmdl</li>
</ul>
</details>

File: pkgs/development/python-modules/pydes/default.nix [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+  version = "unstable-2019-01-08";
+
+  src = fetchFromGitHub {
+    owner = "twhiteman";
+    repo = "pyDes";
+    rev = "e988a5ffc9abb8010fc75dba54904d1c5dbe83db";
+    sha256 = "0sic8wbyk5azb4d4m6zbc96lfqcw8s2pzcv9nric5yqc751613ww";
+  };
+
+  checkPhase = "python test_pydes.py";
```
Comment (dotlambda) [inline pkgs/development/python-modules/pydes/default.nix:14]: You should use `${python.interpreter}`.

File: pkgs/tools/misc/ytmdl/default.nix [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+  };
+
+  postPatch = ''
+    substituteInPlace setup.py \
+      --replace "bs4" "beautifulsoup4"
+  '';
+
+
+  propagatedBuildInputs = [
+    ffmpeg
```
Comment (dotlambda) [inline pkgs/tools/misc/ytmdl/default.nix:21]: I doubt this belongs in `propagatedBuildInputs`. Maybe you can use `makeWrapperArgs` to add it to `$PATH`?

**Pull request merged**